### ISSUE TITLE
perf: calc - doc_setView shouldn't change toolbars

### ISSUE THREAD

**Issue by mmeeks** [CollaboraOnline/online#7428]
Switching view should not of itself fiddle with toolbars or the notebookbar etc. it seems in the trace here:

https://user-images.githubusercontent.com/833656/274579731-f8946ec3-74da-4d50-91fc-518d6c14639c.svg

We have SfxDispatcher::Update_Impl being called on view switching; in every mouse move it updates the toolbar bits, but also in every paint tile, in every input event etc. - eg. the cost of getCommandValues is totally dominated by pointless doc_setView work being done:

![image](https://github.com/CollaboraOnline/online/assets/122848/eddf8bd7-a18d-43d4-b8b4-d2b3032d1fa0)

Quickly adding this up I see around 10% of total CPU cycles spent.

It seems that Document::renderTiles - does a doc_setView - costing 8bn cycles, which then does another doc_setView inside itself costing 11bn+ cycles.

Quite probably this is caused by the outer version not understanding canoncial-views, and creating extra work - nevertheless - perhaps an easy one-liner win to remove the doc_setView in Document::renderTiles (?) or ...

![image](https://github.com/CollaboraOnline/online/assets/122848/4e7a54b5-48d9-4fdf-97eb-cecdf27a4ee0)

Thanks ! =)

**Comment by caolanm** [comment]
This is an odd one.

We do a _loKitDocument->setView in kit/Kit.cpp in renderTiles and locally with multiple connections to a calc document I see no sign of an inner setView from doc_paintPartTile getting called unlike seen above.

Also in the flamegraph I see the WeldedTabbedNotebookbar ctor getting called a load of times, eating most of the doc_setView time, something I don't see locally.

The WeldedTabbedNotebookbar is created from SfxNotebookBar::StateMethod which is called from ScDocShell::GetState. In ScDocShell::GetState SfxNotebookBar::StateMethod is called with bReloadNotebookbar of false, so presumably the creation must be due to the other condition of m_pNotebookBarWeldedWrapper.find(pViewShell) != m_pNotebookBarWeldedWrapper.end(). But we keep that mapping of viewshell of notebookbar until VclDisposeHdl is called.

Locally I see one notebookbar per view, created when a user joins and not torn down to be replaced until that user exits.

I could speculate that the first issue was related to the recently fixed DocumentBroker::sendRequestedTiles but I'm not sure the same could account for the second issue.

**Comment by caolanm** [comment]
Once difference I see from my local case and debugging staging is that locally I get one notebookbar created per user, while in staging it is created, torn down and recreated.

In both cases the browser starts the notebookbar with:

createNotebookbarControl (Control.UIManager.js:442)
initializeSpecializedUI (Control.UIManager.js:289)
beforeAdd (WriterTileLayer.js:24)
addLayer (Layer.js:69)
_onStatusMsg (Socket.js:1384)
_onMessage (Socket.js:1162)
_emitSlurpedEvents (Socket.js:347)
(anonymous) (Socket.js:297)
setTimeout (async)
_queueSlurpEventEmission (Socket.js:295)
_slurpMessage (Socket.js:395)
_emptyQueue (Socket.js:80)
connect (Socket.js:74)
loadDocument (Map.js:321)
(anonymous) (main.js:93)
(anonymous) (main.js:113)

while in the staging case it goes on to reset the notebookbar with:

onRemove (Control.Notebookbar.js:105)
remove (Control.js:71)
removeControl (Control.js:104)
removeNotebookbarUI (Control.UIManager.js:501)
refreshNotebookbar (Control.UIManager.js:460)
refreshUI (Control.Zotero.js:140)
updateUserID (Control.Zotero.js:147)
onUpdateViews (Control.UIManager.js:867)
fire (Events.js:152)
_onViewInfoMsg (CanvasTileLayer.js:2969)
_onMessage (CanvasTileLayer.js:1763)
_onMessage (Socket.js:1197)
_emitSlurpedEvents (Socket.js:347)
(anonymous) (Socket.js:297)
setTimeout (async)
_queueSlurpEventEmission (Socket.js:295)
_slurpMessage (Socket.js:395)
_emptyQueue (Socket.js:80)
connect (Socket.js:74)
loadDocument (Map.js:321)
(anonymous) (main.js:93)
(anonymous) (main.js:113)

and the difference is Zotero support which is off my default locally. updateUserID (Control.Zotero.js:147)

Switching dark/light mode locally is another trigger.

Maybe in the case we are "refreshing" the browser notebookbar we don't have to remove and recreate the "core" one. Or at least not for the a light/dark mode switch.

**Comment by grandinj** [comment]
At sfx2/source/notebookbar/SfxNotebookBar.cxx:380 in  SfxNotebookBar::StateMethod we have code that looks like

     || bReloadNotebookbar || comphelper::LibreOfficeKit::isActive())

which means that in LOK we forcefully recreate the toolbars every time we enter this method. That code came from 

    https://cgit.freedesktop.org/libreoffice/core/commit/?id=004512d87ab70

@eszkadev can you remember why you needed to do it that way?

**Comment by eszkadev** [comment]
@grandinj it looks like we can change that

https://gerrit.libreoffice.org/c/core/+/158872

**Comment by eszkadev** [comment]
I found why it was done that way. Now it doesn't work properly with multiple sessions switching on/off notebookbar.
Bug:
1. Open 2 sessions with notebookbar
2. Switch user B to compact mode
3. Try to use widgets (style previews) in user A - it doesn't work
I have to add cypress test for that...

I fixed that regression and prepared better version:
https://gerrit.libreoffice.org/c/core/+/158936
https://gerrit.libreoffice.org/c/core/+/158937
https://gerrit.libreoffice.org/c/core/+/158939

With above changes I see no impact of `SfxNotebookbar::StateMethod` or `SfxNotebookBar::IsActive` on `doc_setView`. Also it works again with multiple views switching on and off the notebookbar. (It creates notebookbar when session uses it and destroys when is closed, in contrast to original state when every time we called `SfxNotebookbar::StateMethod` it was recreated.
